src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -2,108 +2,1037 @@
+                    ValidateAndTrackHeader(HttpKnownHeaderNames.Upgrade, "websocket", headerName, headerValue, ref foundUpgrade);
+                    ValidateAndTrackHeader(HttpKnownHeaderNames.SecWebSocketAccept, expectedSecWebSocketAccept, headerName, headerValue, ref foundSecWebSocketAccept);
+
+                    // The SecWebSocketProtocol header is optional.  We should only get it with a non-empty value if we requested subprotocols,
+                    // and then it must only be one of the ones we requested.  If we got a subprotocol other than one we requested (or if we
+                    // already got one in a previous header), fail. Otherwise, track which one we got.
+                    if (string.Equals(HttpKnownHeaderNames.SecWebSocketProtocol, headerName, StringComparison.OrdinalIgnoreCase) &&
+                        !string.IsNullOrWhiteSpace(headerValue))
+                    {
+                        string newSubprotocol = options.RequestedSubProtocols.Find(requested => string.Equals(requested, headerValue, StringComparison.OrdinalIgnoreCase));
+                        if (newSubprotocol == null || subprotocol != null)
+                        {
+                            throw new WebSocketException(
+                                WebSocketError.UnsupportedProtocol,
+                                SR.Format(SR.net_WebSockets_AcceptUnsupportedProtocol, string.Join(", ", options.RequestedSubProtocols), subprotocol));
+                        }
+                        subprotocol = newSubprotocol;
+                    }
+                }
+                if (!foundUpgrade || !foundConnection || !foundSecWebSocketAccept)
+                {
+                    throw new WebSocketException(SR.net_webstatus_ConnectFailure);
+                }
+
+                return subprotocol;
+            }
+
+            private static void ValidateAndTrackHeader(
+                string targetHeaderName, string targetHeaderValue, 
+                string foundHeaderName, string foundHeaderValue, 
+                ref bool foundHeader)
+            {
+                bool isTargetHeader = string.Equals(targetHeaderName, foundHeaderName, StringComparison.OrdinalIgnoreCase);
+                if (!foundHeader)
+                {
+                    if (isTargetHeader)
+                    {
+                        if (!string.Equals(targetHeaderValue, foundHeaderValue, StringComparison.OrdinalIgnoreCase))
+                        {
+                            throw new WebSocketException(SR.Format(SR.net_WebSockets_InvalidResponseHeader, targetHeaderName, foundHeaderValue));
+                        }
+                        foundHeader = true;
+                    }
+                }
+                else
+                {
+                    if (isTargetHeader)
+                    {
+                        throw new WebSocketException(SR.Format(SR.net_webstatus_ConnectFailure));
```

Comment:
I notice that there are a lot of places that can cause ConnectFailure; and they're potentially hard to diagnose.  Do we need/want logging in this class?